pkg/conversion/converter.go

```diff
@@ -0,0 +1,231 @@
+
+	return convertedList, nil
+}
+
+func (c *Converter) convert(ctx context.Context, original *unstructured.Unstructured, targetGV schema.GroupVersion) (*unstructured.Unstructured, error) {
+	converted := original.DeepCopy()
+	converted.SetAPIVersion(targetGV.String())
+
+	originalVersion := original.GetObjectKind().GroupVersionKind().Version
+	compiledRules := c.compiledRules[originalVersion]
+	for _, rule := range compiledRules {
+		value, err := evaluateRule(ctx, original, rule)
+		if err != nil {
+			return nil, fmt.Errorf("error converting: %w", err)
+		}
+
+		if err := unstructured.SetNestedField(converted.Object, value, rule.ToFields...); err != nil {
+			return nil, fmt.Errorf("error setting destination field %q: %w", rule.ToPath, err)
+		}
+	}
+
+	annotations := converted.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	preserveAnnotation := annotations[preserveAnnotationKey(targetGV.Version)]
+	if preserveAnnotation != "" {
+		// If we're going to a version that already has the preserve annotation, restore from it
+		m := map[string]interface{}{}
+		if err := json.Unmarshal([]byte(preserveAnnotation), &m); err != nil {
+			return nil, fmt.Errorf("error unmarshaling preserve annotation: %v", err)
+		}
+		if err := restoreFromPreserveMap(m, converted.Object); err != nil {
+			return nil, fmt.Errorf("error processing preserve map: %w", err)
+		}
+
+		// Don't save the preserve annotation
+		delete(annotations, preserveAnnotationKey(targetGV.Version))
+		if len(annotations) == 0 {
+			converted.SetAnnotations(nil)
+		} else {
+			converted.SetAnnotations(annotations)
+		}
+	} else {
+		// Otherwise, store it
+		m, err := generatePreserveMap(original, c.apiConversion.Spec.Conversions)
+		if err != nil {
+			return nil, fmt.Errorf("error preserving fields: %w", err)
+		}
+
+		if len(m) > 0 {
+			encoded, err := json.Marshal(m)
+			if err != nil {
+				return nil, fmt.Errorf("error encoding preserve map: %v", err)
+			} else {
+				annotations[preserveAnnotationKey(originalVersion)] = string(encoded)
+				converted.SetAnnotations(annotations)
+			}
+		}
+	}
+
+	return converted, nil
+}
+
+func preserveAnnotationKey(version string) string {
+	return apisv1alpha1.VersionPreservationAnnotationKeyPrefix + version
+}
+
+func evaluateRule(ctx context.Context, original *unstructured.Unstructured, rule *CompiledRule) (interface{}, error) {
+	fromValue, exists, err := unstructured.NestedFieldNoCopy(original.Object, rule.FromFields...)
+	if err != nil {
+		return nil, fmt.Errorf("error getting source field %q: %w", rule.FromPath, err)
+	}
+	if !exists {
+		return nil, nil
+	}
+
+	if rule.Program == nil {
+		return fromValue, nil
+	}
+
+	bindings := map[string]interface{}{
+		"self": fromValue,
+	}
+	activation, err := interpreter.NewActivation(bindings)
+	if err != nil {
+		return nil, fmt.Errorf("error creating CEL interpreter activation: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
```

Comment:
Should this be configurable? Is it necessary for the first iteration of this API?

Reply:
@sttts suggested putting a timeout in here instead of `context.TODO()`. 1 second per rule execution seems super generous (probably too much, but maybe CI needs it). I also considered an overall timeout in `convert` and passing that in here. Maybe that's better?

Not sure it should be configurable. This is to protect the apiserver from runaway conversions.

Reply:
Ah, this is _per-rule_ which yeah makes it very generous indeed; making it global might make sense